### ISSUE TITLE
App crash after first configuration modification in webUI

### ISSUE THREAD

**Issue by mrjk** [tiredofit/docker-backuppc#15]
### Summary

App crash after first configuration modification in webUI, especially when you add a new host.

### Steps to reproduce

* Start docker compose
* Get into webui
* Try to add an host
* Press the save button, the app is now crashed

### What is the expected *correct* behavior?

It should not crash the app.

### Relevant logs and/or screenshots

After a fresh install, the first time I want to edit configuration via the WebUI and save it, I get this error:
```
Software error:

Not an ARRAY reference at /usr/local/BackupPC/lib/BackupPC/CGI/Lib.pm line 468.

For help, please send mail to this site's webmaster, giving this error message and the time and date of the error. 
```

When I diff the two config file, I see it changed in a different structure, mainly replacing `[` by `)`, like this diff shows:
```
 diff -u conf/etc/config.pl conf/etc/config.pl.old 
--- conf/etc/config.pl	2022-12-20 00:34:51.435645568 +0000
+++ conf/etc/config.pl.old	2022-12-20 00:11:30.109434305 +0000
@@ -113,7 +113,7 @@
 # you want BackupPC_nightly to run (eg: when you don't expect a lot
 # of regular backups to run).
 #
-$Conf{WakeupSchedule} = (
+$Conf{WakeupSchedule} = [
   1,
   2,
   3,
@@ -137,7 +137,7 @@
   21,
   22,
   23
-);
+];
 
```
<details>
<summary>After webUI config edit</summary>
```
backuppc-backuppc-1  |                                                                        ,---.
backuppc-backuppc-1  | ,--------.,--.                 ,--.            ,---.    ,--. ,--------.|   |
backuppc-backuppc-1  | '--.  .--'`--',--.--. ,---.  ,-|  |     ,---. /  .-'    |  | '--.  .--'|  .'
backuppc-backuppc-1  |    |  |   ,--.|  .--'| .-. :' .-. |    | .-. ||  `-,    |  |    |  |   |  |
backuppc-backuppc-1  |    |  |   |  ||  |   \   --.\ `-' |    ' '-' '|  .-'    |  |.--.|  |   `--'
backuppc-backuppc-1  |    `--'   `--'`--'    `----' `---'      `---' `--'      `--''--'`--'   .--.
backuppc-backuppc-1  |                                                                        '--'
backuppc-backuppc-1  | Image:  tiredofit/backuppc | Version  6.0.0 Type 'image_changelog' for details
backuppc-backuppc-1  | Repository/Issues/Support:  https://github.com/tiredofit/docker-backuppc/
backuppc-backuppc-1  |            Sponsor me for development and upkeep: https://www.tiredofit.ca/sponsor
backuppc-backuppc-1  |
backuppc-backuppc-1  | 2022-12-19.19:42:42 [NOTICE] ** [monitoring] Container configured for monitoring with 'zabbix modern'
backuppc-backuppc-1  | 2022-12-19.19:42:42 [NOTICE] ** [scheduling] Container configured for scheduled tasks with 'cron'
backuppc-backuppc-1  | 2022-12-19.19:42:43 [NOTICE] ** [messaging] Container configured to route mail via SMTP to 'postfix-relay'
backuppc-backuppc-1  | 2022-12-19.19:42:44 [INFO] ** [backuppc] BackupPC 4.4.0 initialization complete, now starting web server on port 80
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [fcgiwrap] [1] Starting fcgiwrap
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [nginx] [1] Starting nginx 1.23.2
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [backuppc] [1] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [monitoring] [1] Starting Zabbix Agent (modern) 6.2.4
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [scheduling] [1] Starting cron
backuppc-backuppc-1  | 2022-12-19.19:47:32 [STARTING] ** [backuppc] [2] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:33 [STARTING] ** [backuppc] [3] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:34 [STARTING] ** [backuppc] [4] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:35 [STARTING] ** [backuppc] [5] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:36 [STARTING] ** [backuppc] [6] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:37 [STARTING] ** [backuppc] [7] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:38 [STARTING] ** [backuppc] [8] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:39 [STARTING] ** [backuppc] [9] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:40 [STARTING] ** [backuppc] [10] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:41 [STARTING] ** [backuppc] [11] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:42 [STARTING] ** [backuppc] [12] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:43 [STARTING] ** [backuppc] [13] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:44 [STARTING] ** [backuppc] [14] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:45 [STARTING] ** [backuppc] [15] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:46 [STARTING] ** [backuppc] [16] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:47 [STARTING] ** [backuppc] [17] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:48 [STARTING] ** [backuppc] [18] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:49 [STARTING] ** [backuppc] [19] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:50 [STARTING] ** [backuppc] [20] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:51 [STARTING] ** [backuppc] [21] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:52 [STARTING] ** [backuppc] [22] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:53 [STARTING] ** [backuppc] [23] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:54 [STARTING] ** [backuppc] [24] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:55 [STARTING] ** [backuppc] [25] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:56 [STARTING] ** [backuppc] [26] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:57 [STARTING] ** [backuppc] [27] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:58 [STARTING] ** [backuppc] [28] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:47:59 [STARTING] ** [backuppc] [29] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:00 [STARTING] ** [backuppc] [30] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:01 [STARTING] ** [backuppc] [31] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:02 [STARTING] ** [backuppc] [32] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:03 [STARTING] ** [backuppc] [33] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:04 [STARTING] ** [backuppc] [34] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:05 [STARTING] ** [backuppc] [35] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:06 [STARTING] ** [backuppc] [36] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:07 [STARTING] ** [backuppc] [37] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:08 [STARTING] ** [backuppc] [38] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:09 [STARTING] ** [backuppc] [39] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:10 [STARTING] ** [backuppc] [40] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:11 [STARTING] ** [backuppc] [41] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:12 [STARTING] ** [backuppc] [42] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:13 [STARTING] ** [backuppc] [43] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:14 [STARTING] ** [backuppc] [44] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:15 [STARTING] ** [backuppc] [45] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:16 [STARTING] ** [backuppc] [46] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:17 [STARTING] ** [backuppc] [47] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:18 [STARTING] ** [backuppc] [48] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:19 [STARTING] ** [backuppc] [49] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:20 [STARTING] ** [backuppc] [50] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:48:21 [ERROR] ** [backuppc] POTENTIAL RUNWAY DETECTECTED: Disabling 20-backuppc service because it has tried restarting '50' times
```
</details>

### Environment

From inside the container:
```
[tiredofit/backuppc:6.0.0 19:43:31 /data] $ su -s /bin/bash  backuppc
[tiredofit/backuppc:6.0.0 19:43:37 /data] $ whoami 
backuppc
[tiredofit/backuppc:6.0.0 19:43:40 /data] $ /usr/local/BackupPC/bin/BackupPC --help
/usr/local/BackupPC/bin/BackupPC version [unknown] calling Getopt::Std::getopts (version 1.13 [paranoid]),
running under Perl version 5.36.0.

Usage: BackupPC [-OPTIONS [-MORE_OPTIONS]] [--] [PROGRAM_ARG1 ...]

The following single-character options are accepted:
	Boolean (without arguments): -d

Options may be merged together.  -- stops processing of options.
  [Now continuing due to backward compatibility and excessive paranoia.
   See 'perldoc Getopt::Std' about $Getopt::Std::STANDARD_HELP_VERSION.]
2022-12-19 19:43:56 Another BackupPC is running (pid 1525); quitting...
```

- Image version / tag:  tiredofit/backuppc:6.0.0
- Host OS: Debian 11

<details>
<summary>Before editing config</summary>
```
backuppc-backuppc-1  |                                                                        ,---.
backuppc-backuppc-1  | ,--------.,--.                 ,--.            ,---.    ,--. ,--------.|   |   
backuppc-backuppc-1  | '--.  .--'`--',--.--. ,---.  ,-|  |     ,---. /  .-'    |  | '--.  .--'|  .'  
backuppc-backuppc-1  |    |  |   ,--.|  .--'| .-. :' .-. |    | .-. ||  `-,    |  |    |  |   |  |
backuppc-backuppc-1  |    |  |   |  ||  |   \   --.\ `-' |    ' '-' '|  .-'    |  |.--.|  |   `--'
backuppc-backuppc-1  |    `--'   `--'`--'    `----' `---'      `---' `--'      `--''--'`--'   .--.
backuppc-backuppc-1  |                                                                        '--'
backuppc-backuppc-1  | Image:  tiredofit/backuppc | Version  6.0.0 Type 'image_changelog' for details
backuppc-backuppc-1  | Repository/Issues/Support:  https://github.com/tiredofit/docker-backuppc/
backuppc-backuppc-1  |            Sponsor me for development and upkeep: https://www.tiredofit.ca/sponsor
backuppc-backuppc-1  |   
backuppc-backuppc-1  | 2022-12-19.19:42:42 [NOTICE] ** [monitoring] Container configured for monitoring with 'zabbix modern'
backuppc-backuppc-1  | 2022-12-19.19:42:42 [NOTICE] ** [scheduling] Container configured for scheduled tasks with 'cron'
backuppc-backuppc-1  | 2022-12-19.19:42:43 [NOTICE] ** [messaging] Container configured to route mail via SMTP to 'postfix-relay'
backuppc-backuppc-1  | 2022-12-19.19:42:44 [INFO] ** [backuppc] BackupPC 4.4.0 initialization complete, now starting web server on port 80
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [fcgiwrap] [1] Starting fcgiwrap
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [nginx] [1] Starting nginx 1.23.2
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [backuppc] [1] Starting BackupPC 4.4.0
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [monitoring] [1] Starting Zabbix Agent (modern) 6.2.4
backuppc-backuppc-1  | 2022-12-19.19:42:44 [STARTING] ** [scheduling] [1] Starting cron
```
</details>

### Possible fixes

A possible workaround is to NOT use the web ui to edit the config OR to fix the file with a text editor after each UI modifcation.

I'm not a perl guy, but I'm wondering if the current version of perl (5.36.0) still support backuppc, wich is basically compatible since perl 5.8 (2002!).  Should we try with a older perl version ?

In the end, I'm not sure anymore I want to use a software that is not really maintened anymore. It seems you are/were an heavy user of backuppc, do you now use something else ?

**Comment by mrjk** [comment]
I tried with an older version, everything works well with `tiredofit/backuppc:5.3.10`. I'll use this one by the time to evaluate backuppc :)

**Comment by tiredofit** [comment]
Interesting. #13 also spoke of compatibility issues. I think this for the time being warrants rolling back to a different base image (Alpine 3.16 - still very much supported) which uses perl `5.34.1`. 

I've tagged `tiredofit/backuppc:6.0.1` as to be this rolled back version.

On backups - I've used BackupPC since 2004 ish at many sites, and it does the job. It uses a programming language which while not as popular as it once was, again, does the job. It's one of the few perl applications I still use today, the other being [this](https://github.com/tiredofit/docker-lemonldapl).

Re Your backup alternative question:
I'm now testing in parallel to BackupPC a client and (sorta kinda) server based system called [Kopia](https://kopia.io) to which I've made starts of an [image ](https://github.com/tiredofit/docker-kopia). It has a central "Repository Server" to send backups to, with unique usernames and passwords for each host connecting to it, which then encrypts and sends it to a destination, in my case I am sending it to a S3 bucket (Wasabi) for far less of what I pay for one of my BackupPC systems. I've only been running for 45 days, and backups and restores seem to be working just fine. Monitoring is non existent (It outputs Prometheus metrics, but that is not monitoring) and logging is embarrassingly bad with logs not being saved to filesystem if setup via the UI. I wanted to test the durability of it before going any further on it, I have 3 more months to investigate before I make a decision on what I'd like to do in the future. It is very very fast, has parity, encryption and compression out of the box along with deduplication and the ability to connect to any remote filesystem that Rclone supports.  Other alternatives seem to be Restic which doesn't seem to have a central server requiring me to manage multiple hosts regularly along with separate s3 credentials, and Duplicati, although performance is not great for all of them compared to Kopia. Apples and Oranges for sure.

I still feel BackupPC is a great system and serves a purpose for specific environments, I have half a dozen instances still running and will continue to do so on some areas that have limited bandwidth to move to "cloud" styled storage backups.

Let me know if the base image change gets you working again.

**Comment by mrjk** [comment]
Ok, I tried out the version 6.0.1, and it fixed the issue :) Thank you very much. I believed as well the app was failing to start because of some permissions errors when `USER_BACKUPPC` and/or `GROUP_BACKUPPC` were unset, but unable to reproduce it. 

On backups:
I've toyed with lemonldap as well, it was painful to deploy, especially in a docker environment. I should have tested with your images first. I eventually took another tool in the end.
I'll definitely look for Kopia, seems an interesting alternative. I need to stick to the server model on my case, as I can't afford to install more things that ssh and rsync on most the devices I want to backup. It's pretty hard to find an opensource centralized backup solution with a not that bad UI that allows you to trigger jobs and give you a feedback like: yeah, I got you covered back, you are now allowed to safely mess this old production server \o/ For now, I'll stick with backuppc, but when I'll have more free time, I'll give it a try.

Thanks for your feedback, and your prompt 6.0.1 patch ;)
